sw/device/boot_rom2/final_jump.S

```diff
@@ -0,0 +1,150 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+  /**
+   * Linker-provided symbols for stack layout information.
+   *
+   * We assume that the stack is a word-aligned array of words,
+   * so we can use word loads and stores on it exclusively.
+   */
+  .extern _stack_start
+  .extern _stack_size
+
+  /**
+   * Linker-provided symbols for .bss and .data regions.
+   *
+   * These are all known to be word-aligned.
+   */
+  .extern _bss_start
+  .extern _bss_end
+  .extern _data_start
+  .extern _data_end
+
+// noreturn void _final_jump_asm(entrypoint_t entry, void *next_stack_start);
+_final_jump_asm:
+  .globl _final_jump_asm
+  // Since this function is called from C, the registers |a0| and |a1|
+  // correspond to the function arguments |entry| and |next_stack_start|,
+  // respectively.
+
+  // Bleach the ROM stack, which we no longer need for anything.
+  // Here, |t0| is the current word being zeroed, while |t1| is
+  // the number of bytes left to zero.
+  //
+  // This operation is "done in reverse", from the bottom of the stack
+  // to the top. As such, |t0| decreases with each write.
+  la   t0, _stack_start
+  la   t1, _stack_size
+stack_bleach:
+  addi t0, t0, -4
+  addi t1, t1, -4
+  sw   zero, 0(t0)
+  bnez t1, stack_bleach
+
+  // Bleach the .bss section. |t0| is the current word being zeroed,
+  // while |t1| is the end pointer.
+  //
+  // This operation is done in forward order.
+  la  t0, _bss_start
+  la  t1, _bss_end
+  bge t0, t1, bss_bleach_end
+bss_bleach:
+  sw   zero, 0(t0)
+  addi t0, t0, 4
+  ble  t0, t1, bss_bleach
+bss_bleach_end:
+
+  // Do the exact same bleach operation to the .data section.
+  la  t0, _data_start
+  la  t1, _data_end
+  bge t0, t1, data_bleach_end
+data_bleach:
+  sw   zero, 0(t0)
+  addi t0, t0, 4
+  ble  t0, t1, data_bleach
+data_bleach_end:
```

Comment:
This is reasonable because it's the new image's responsibility to load its own `.idata` section into `.data`, right?